ComponentKit/DataSources/CKCollectionViewTransactionalDataSource.mm

```diff
@@ -0,0 +1,146 @@
+    [collectionView moveItemAtIndexPath:from toIndexPath:to];
+  }
+  [collectionView insertSections:changes.insertedSections];
+  [collectionView insertItemsAtIndexPaths:[changes.insertedIndexPaths allObjects]];
+}
+
+#pragma mark - CKTransactionalComponentDataSourceListener
+
+- (void)transactionalComponentDataSource:(CKTransactionalComponentDataSource *)dataSource
+                  didModifyPreviousState:(CKTransactionalComponentDataSourceState *)previousState
+                       byApplyingChanges:(CKTransactionalComponentDataSourceAppliedChanges *)changes
+{
+  [_collectionView performBatchUpdates:^{
+    applyChangesToCollectionView(changes, _collectionView);
+    // Detach all the component layouts for items being deleted
+    [self _detachComponentLayoutForRemovedItemsAtIndexPaths:[changes removedIndexPaths]
+                                                    inState:previousState];
+    // Update current state
+    _currentState = [_componentDataSource state];
+  } completion:NULL];
+}
+
+- (void)_detachComponentLayoutForRemovedItemsAtIndexPaths:(NSSet *)removedIndexPaths
+                                                  inState:(CKTransactionalComponentDataSourceState *)state
+{
+  for (NSIndexPath *indexPath in removedIndexPaths) {
+    CKComponentScopeRootIdentifier identifier = [[[state objectAtIndexPath:indexPath] scopeRoot] globalIdentifier];
+    [_attachController detachComponentLayoutWithScopeIdentifier:identifier];
+  }
+}
+
+#pragma mark - State
+
+- (CGSize)sizeForItemAtIndexPath:(NSIndexPath *)indexPath
+{
+  return [_currentState objectAtIndexPath:indexPath].layout.size;
+}
+
+#pragma mark - Reload
+
+- (void)reloadWithMode:(CKTransactionalComponentDataSourceMode)mode
+              userInfo:(NSDictionary *)userInfo
+{
+  [_componentDataSource reloadWithMode:mode userInfo:userInfo];
+}
+
+- (void)updateConfiguration:(CKTransactionalComponentDataSourceConfiguration *)configuration
+                       mode:(CKTransactionalComponentDataSourceMode)mode
+                   userInfo:(NSDictionary *)userInfo
+{
+  [_componentDataSource updateConfiguration:configuration mode:mode userInfo:userInfo];
+}
+
+#pragma mark - UICollectionViewDataSource
+
+static NSString *const kReuseIdentifier = @"com.component_kit.collection_view_data_source.cell";
+
+- (UICollectionViewCell *)collectionView:(UICollectionView *)collectionView cellForItemAtIndexPath:(NSIndexPath *)indexPath
+{
+  CKCollectionViewDataSourceCell *cell = [_collectionView dequeueReusableCellWithReuseIdentifier:kReuseIdentifier forIndexPath:indexPath];
+  CKTransactionalComponentDataSourceItem *item = [_currentState objectAtIndexPath:indexPath];
+  [_attachController attachComponentLayout:item.layout withScopeIdentifier:item.scopeRoot.globalIdentifier toView:cell.rootView];
+  return cell;
+}
+
+- (NSInteger)numberOfSectionsInCollectionView:(UICollectionView *)collectionView
+{
+  return _currentState ? [_currentState numberOfSections] : 0;
```

Comment:
Yeah, this is on purpose to make it explicit that 0 is a default value we choose. Can change if you want, will have to fix the way we deal with the initial state anyway.